File: 1046.java
import java.util.*;

class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> heap = new PriorityQueue<>((x, y) -> y - x);
        for (int i : stones) {
            heap.offer(i);
        }
        while (heap.size() > 1) {
            int result = heap.poll() - heap.poll();
            if (result != 0) {
                heap.offer(result);
            }
        }
        return heap.size() == 0 ? 0 : heap.peek();
    }
}

File: Python/1046.py
import heapq
from typing import List


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        heap = [-x for x in stones]
        heapq.heapify(heap)

        while len(heap) >= 2:
            s1 = -1 * heapq.heappop(heap)
            s2 = -1 * heapq.heappop(heap)
            if s1 == s2:
                continue
            else:
                heapq.heappush(heap, -1 * abs(s1 - s2))

        return 0 if len(heap) == 0 else -1 * heap[0]
